src/util/fade_helper.cpp

```diff
@@ -0,0 +1,74 @@
+//  SuperTux
+//  Copyright (C) 2021 A. Semphris <semphris@protonmail.com>
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#include "util/fade_helper.hpp"
+
+FadeHelper::FadeHelper(float time, float target_value,
+                       float start_value, easing ease) :
+  m_value(nullptr),
+  m_start(start_value),
+  m_target(target_value),
+  m_time(0.f),
+  m_total_time(time),
+  m_ease(ease)
+{
+}
+
+FadeHelper::FadeHelper(float* value, float time,
+                       float target_value, easing ease) :
+  m_value(value),
+  m_start(*value),
+  m_target(target_value),
+  m_time(0.f),
+  m_total_time(time),
+  m_ease(ease)
+{
+}
+
+float
+FadeHelper::update(float dt_sec)
+{
+  m_time += dt_sec;
+
+  if (completed())
+  {
+    m_time = m_total_time;
+    if (m_value)
+      *m_value = m_target;
+    return m_target;
+  }
+  else
```

Comment:
Since you return'ed from the last if, the else statement is redundant. You can reduce the level of nesting here.

Reply:
I'm not sure to understand... The value computed for each `return` statement is different between the if and the else.

Do you suggest I put a `float return_value;` that I set in both the if and the else, and then have only 1 return statement after the whole if/else?

Reply:
Fixed :+1: